File: frontend/src/types/document.ts
export type DocumentImportAction = 'created' | 'version_created' | 'unchanged'
export type DocumentParseStatus = 'pending' | 'processing' | 'succeeded' | 'failed'
export type DocumentIndexStatus = 'pending' | 'processing' | 'succeeded' | 'failed'

export interface DocumentVersion {
  id: string
  version_number: number
  content_hash: string
  file_size: number
  mime_type: string | null
  extension: string
  created_at: string
  parse_status: DocumentParseStatus
  parser_name: string | null
  parser_version: string | null
  chunk_count: number
  parse_started_at: string | null
  parsed_at: string | null
  last_parse_attempt_at: string | null
  parse_error_code: string | null
  parse_error_message: string | null
  index_status: DocumentIndexStatus
  active_index_generation: string | null
  index_started_at: string | null
  indexed_at: string | null
  last_index_attempt_at: string | null
  indexed_chunk_count: number
  embedding_model: string | null
  embedding_dimension: number | null
  index_error_code: string | null
  index_error_message: string | null
}

export interface DocumentItem {
  id: string
  knowledge_base_id: string
  name: string
  source_type: string
  created_at: string
  updated_at: string
  version_count: number
  latest_version: DocumentVersion
}

export interface DocumentListResponse {
  items: DocumentItem[]
  total: number
  offset: number
  limit: number
}

export interface DocumentImportResponse {
  import_action: DocumentImportAction
  parsing_queued: boolean
  document: DocumentItem
}

export interface DocumentParseStatusResponse {
  version_id: string
  parse_status: DocumentParseStatus
  parser_name: string | null
  parser_version: string | null
  chunk_count: number
  parse_started_at: string | null
  parsed_at: string | null
  last_parse_attempt_at: string | null
  parse_error_code: string | null
  parse_error_message: string | null
}

export interface DocumentParseRequestResponse {
  queued: boolean
  version: DocumentParseStatusResponse
}

export interface DocumentChunk {
  id: string
  chunk_index: number
  content: string
  content_hash: string
  char_count: number
  page_number: number | null
  start_line: number | null
  end_line: number | null
  section_title: string | null
  chunk_type: string
  language: string | null
  created_at: string
}

export interface DocumentChunkListResponse {
  items: DocumentChunk[]
  total: number
  offset: number
  limit: number
  version: DocumentParseStatusResponse
}

export interface DocumentIndexStatusResponse {
  version_id: string
  index_status: DocumentIndexStatus
  active_index_generation: string | null
  index_started_at: string | null
  indexed_at: string | null
  last_index_attempt_at: string | null
  indexed_chunk_count: number
  embedding_model: string | null
  embedding_dimension: number | null
  index_error_code: string | null
  index_error_message: string | null
}

export interface DocumentIndexRequestResponse {
  queued: boolean
  version: DocumentIndexStatusResponse
}

export interface SemanticSearchResult {
  score: number
  content: string
  knowledge_base_id: string
  document_id: string
  document_version_id: string
  chunk_id: string
  index_generation: string
  document_name: string
  version_number: number
  chunk_index: number
  content_hash: string
  chunk_type: string
  language: string | null
  section_title: string | null
  page_number: number | null
  start_line: number | null
  end_line: number | null
}

export interface SemanticSearchResponse {
  items: SemanticSearchResult[]
}

File: backend/app/models/document.py
from datetime import datetime
from typing import TYPE_CHECKING
from uuid import UUID, uuid4

from sqlalchemy import (
    CHAR,
    BigInteger,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
    Uuid,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base

if TYPE_CHECKING:
    from app.models.knowledge_base import KnowledgeBase


class Document(Base):
    __tablename__ = "documents"
    __table_args__ = (
        UniqueConstraint(
            "knowledge_base_id",
            "normalized_name",
            name="uq_documents_knowledge_base_normalized_name",
        ),
        Index("ix_documents_knowledge_base_id", "knowledge_base_id"),
    )

    id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    knowledge_base_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey(
            "knowledge_bases.id", name="fk_documents_knowledge_base_id", ondelete="RESTRICT"
        ),
        nullable=False,
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    normalized_name: Mapped[str] = mapped_column(String(255), nullable=False)
    source_type: Mapped[str] = mapped_column(
        String(32), nullable=False, default="upload", server_default="upload"
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now()
    )

    knowledge_base: Mapped["KnowledgeBase"] = relationship(back_populates="documents", lazy="raise")
    versions: Mapped[list["DocumentVersion"]] = relationship(
        back_populates="document",
        cascade="all, delete-orphan",
        passive_deletes=True,
        lazy="raise",
    )


class DocumentVersion(Base):
    __tablename__ = "document_versions"
    __table_args__ = (
        UniqueConstraint(
            "document_id",
            "version_number",
            name="uq_document_versions_document_version",
        ),
        CheckConstraint("version_number > 0", name="ck_document_versions_version_positive"),
        CheckConstraint("file_size > 0", name="ck_document_versions_file_size_positive"),
        CheckConstraint("char_length(content_hash) = 64", name="ck_document_versions_hash_length"),
        CheckConstraint("chunk_count >= 0", name="ck_document_versions_chunk_count_nonnegative"),
        CheckConstraint(
            "parse_status IN ('pending', 'processing', 'succeeded', 'failed')",
            name="ck_document_versions_parse_status",
        ),
        Index("ix_document_versions_document_id", "document_id"),
    )

    id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    document_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey(
            "documents.id",
            name="fk_document_versions_document_id",
            ondelete="CASCADE",
        ),
        nullable=False,
    )
    version_number: Mapped[int] = mapped_column(Integer, nullable=False)
    content_hash: Mapped[str] = mapped_column(CHAR(64), nullable=False)
    file_size: Mapped[int] = mapped_column(BigInteger, nullable=False)
    mime_type: Mapped[str | None] = mapped_column(String(255), nullable=True)
    extension: Mapped[str] = mapped_column(String(32), nullable=False)
    storage_path: Mapped[str] = mapped_column(Text, nullable=False)
    parse_status: Mapped[str] = mapped_column(
        String(32), nullable=False, default="pending", server_default="pending"
    )
    parser_name: Mapped[str | None] = mapped_column(String(64), nullable=True)
    parser_version: Mapped[str | None] = mapped_column(String(32), nullable=True)
    chunk_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0, server_default="0")
    parse_started_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    parsed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    last_parse_attempt_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    parse_error_code: Mapped[str | None] = mapped_column(String(64), nullable=True)
    parse_error_message: Mapped[str | None] = mapped_column(String(500), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )

    document: Mapped[Document] = relationship(back_populates="versions", lazy="raise")
    chunks: Mapped[list["DocumentChunk"]] = relationship(
        back_populates="document_version",
        cascade="all, delete-orphan",
        passive_deletes=True,
        lazy="raise",
    )


class DocumentChunk(Base):
    __tablename__ = "document_chunks"
    __table_args__ = (
        UniqueConstraint(
            "document_version_id",
            "chunk_index",
            name="uq_document_chunks_version_index",
        ),
        CheckConstraint("chunk_index >= 0", name="ck_document_chunks_index_nonnegative"),
        CheckConstraint("char_count > 0", name="ck_document_chunks_char_count_positive"),
        CheckConstraint("char_length(content_hash) = 64", name="ck_document_chunks_hash_length"),
        CheckConstraint(
            "page_number IS NULL OR page_number > 0", name="ck_document_chunks_page_positive"
        ),
        CheckConstraint(
            "start_line IS NULL OR start_line > 0", name="ck_document_chunks_start_line_positive"
        ),
        CheckConstraint(
            "end_line IS NULL OR end_line > 0", name="ck_document_chunks_end_line_positive"
        ),
        CheckConstraint(
            "(start_line IS NULL AND end_line IS NULL) OR "
            "(start_line IS NOT NULL AND end_line IS NOT NULL)",
            name="ck_document_chunks_line_pair",
        ),
        CheckConstraint(
            "start_line IS NULL OR end_line >= start_line",
            name="ck_document_chunks_line_order",
        ),
        Index("ix_document_chunks_document_version_id", "document_version_id"),
    )

    id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    document_version_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey(
            "document_versions.id",
            name="fk_document_chunks_document_version_id",
            ondelete="CASCADE",
        ),
        nullable=False,
    )
    chunk_index: Mapped[int] = mapped_column(Integer, nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    content_hash: Mapped[str] = mapped_column(CHAR(64), nullable=False)
    char_count: Mapped[int] = mapped_column(Integer, nullable=False)
    page_number: Mapped[int | None] = mapped_column(Integer, nullable=True)
    start_line: Mapped[int | None] = mapped_column(Integer, nullable=True)
    end_line: Mapped[int | None] = mapped_column(Integer, nullable=True)
    section_title: Mapped[str | None] = mapped_column(String(500), nullable=True)
    chunk_type: Mapped[str] = mapped_column(String(32), nullable=False)
    language: Mapped[str | None] = mapped_column(String(32), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )

    document_version: Mapped[DocumentVersion] = relationship(back_populates="chunks", lazy="raise")
